servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AmbSingles.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright © 2020 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.api;
+
+import io.servicetalk.concurrent.Cancellable;
+import io.servicetalk.concurrent.SingleSource.Subscriber;
+import io.servicetalk.concurrent.internal.DelayedCancellable;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import javax.annotation.Nullable;
+
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.atomic.AtomicIntegerFieldUpdater.newUpdater;
+
+final class AmbSingles<T> extends Single<T> {
+    private final Single<? extends T>[] singles;
+
+    @SafeVarargs
+    AmbSingles(final Single<? extends T>... singles) {
+        for (Single<? extends T> single : singles) {
+            requireNonNull(single);
+        }
+        this.singles = singles;
+    }
+
+    AmbSingles(final Iterable<Single<? extends T>> singles) {
+        List<Single<? extends T>> allSingles = new ArrayList<>();
+        for (Single<? extends T> single : singles) {
+            allSingles.add(requireNonNull(single));
+        }
+        @SuppressWarnings({"unchecked", "SuspiciousToArrayCall"})
+        Single<? extends T>[] singlesArr = (Single<? extends T>[]) allSingles.toArray(new Single[0]);
+        this.singles = singlesArr;
+    }
+
+    @Override
+    protected void handleSubscribe(final Subscriber<? super T> subscriber) {
+        final Cancellable[] cancellables = new Cancellable[singles.length];
```

Comment:
consider most/all of the allocation/invocations in `try/catch` to avoid letting exceptions escape. we also have more context here as to when `onSubscribe` is invoked to avoid letting the exception bubble up and invoking it multiple times.

Reply:
Good point about `onSubscribe()`. I will handle exceptions thrown from `onSubscribe()` but will leave the allocations outside `try` as it will further complicate control-flow and we expect `Single#subscribe()` to catch and invoke error on `Subscriber` anyways.